File: apps/api/app/schemas/asset.py
"""
Schemas Pydantic para Bienes / Activos Fijos e Inventario Físico (Fase 2).
"""
import uuid
from datetime import date, datetime
from pydantic import BaseModel, ConfigDict
from app.schemas.person import PersonaRead


# ── Movimiento de Bien (Trazabilidad) ─────────────────────────────────────

class BienMovimientoBase(BaseModel):
    bien_id: uuid.UUID
    espacio_origen_id: uuid.UUID | None = None
    espacio_destino_id: uuid.UUID | None = None
    persona_responsable_id: uuid.UUID | None = None
    motivo: str | None = None


class BienMovimientoCreate(BienMovimientoBase):
    pass


class BienMovimientoRead(BienMovimientoBase):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    fecha_traslado: datetime
    espacio_origen_codigo: str | None = None
    espacio_destino_codigo: str | None = None
    persona_responsable_nombre: str | None = None
    created_at: datetime


# ── Bien / Activo Fijo ───────────────────────────────────────────────────

class BienBase(BaseModel):
    codigo_patrimonial: str
    nombre: str
    categoria: str = "MOBILIARIO"  # MOBILIARIO, TI_COMPUTO, CLIMATIZACION, LABORATORIO, AUDIOVISUAL, SEGURIDAD, OTRO
    marca: str | None = None
    modelo: str | None = None
    numero_serie: str | None = None
    estado_operativo: str = "OPERATIVO"  # OPERATIVO, EN_MANTENCION, EN_REPARACION, DE_BAJA, EN_BODEGA
    valor_compra: float | None = None
    fecha_adquisicion: date | None = None
    fecha_garantia: date | None = None
    espacio_id: uuid.UUID | None = None
    custodio_id: uuid.UUID | None = None
    pos_x: float | None = None
    pos_y: float | None = None
    metadata_extra: str | None = None


class BienCreate(BienBase):
    pass


class BienUpdate(BaseModel):
    codigo_patrimonial: str | None = None
    nombre: str | None = None
    categoria: str | None = None
    marca: str | None = None
    modelo: str | None = None
    numero_serie: str | None = None
    estado_operativo: str | None = None
    valor_compra: float | None = None
    fecha_adquisicion: date | None = None
    fecha_garantia: date | None = None
    espacio_id: uuid.UUID | None = None
    custodio_id: uuid.UUID | None = None
    pos_x: float | None = None
    pos_y: float | None = None
    metadata_extra: str | None = None


class BienPositionUpdate(BaseModel):
    pos_x: float
    pos_y: float


class BienTransferRequest(BaseModel):
    espacio_destino_id: uuid.UUID
    persona_responsable_id: uuid.UUID | None = None
    motivo: str | None = "Reasignación de recinto"


class BienRead(BienBase):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    created_at: datetime
    updated_at: datetime


class BienWithDetailsRead(BienRead):
    espacio_codigo: str | None = None
    espacio_nombre: str | None = None
    piso_id: uuid.UUID | None = None
    piso_numero: int | None = None
    edificio_nombre: str | None = None
    sede_nombre: str | None = None
    custodio_nombre: str | None = None
    custodio_cargo: str | None = None
    movimientos: list[BienMovimientoRead] = []


class AssetSummaryStats(BaseModel):
    total_bienes: int
    bienes_operativos: int
    bienes_en_mantencion: int
    bienes_de_baja: int
    por_categoria: dict[str, int] = {}

File: packages/shared-types/src/entities/asset.ts
export type AssetCategory =
  | "MOBILIARIO"
  | "TI_COMPUTO"
  | "CLIMATIZACION"
  | "LABORATORIO"
  | "AUDIOVISUAL"
  | "SEGURIDAD"
  | "OTRO";

export type AssetOperationalStatus =
  | "OPERATIVO"
  | "EN_MANTENCION"
  | "EN_REPARACION"
  | "DE_BAJA"
  | "EN_BODEGA";

export interface BienMovimiento {
  id: string;
  bien_id: string;
  espacio_origen_id?: string | null;
  espacio_destino_id?: string | null;
  persona_responsable_id?: string | null;
  fecha_traslado: string;
  motivo?: string | null;
  espacio_origen_codigo?: string | null;
  espacio_destino_codigo?: string | null;
  persona_responsable_nombre?: string | null;
  created_at: string;
}

export interface Bien {
  id: string;
  codigo_patrimonial: string;
  nombre: string;
  categoria: AssetCategory | string;
  marca?: string | null;
  modelo?: string | null;
  numero_serie?: string | null;
  estado_operativo: AssetOperationalStatus | string;
  valor_compra?: number | null;
  fecha_adquisicion?: string | null;
  fecha_garantia?: string | null;
  espacio_id?: string | null;
  custodio_id?: string | null;
  pos_x?: number | null;
  pos_y?: number | null;
  metadata_extra?: string | null;
  espacio_codigo?: string | null;
  espacio_nombre?: string | null;
  piso_id?: string | null;
  piso_numero?: number | null;
  edificio_nombre?: string | null;
  sede_nombre?: string | null;
  custodio_nombre?: string | null;
  custodio_cargo?: string | null;
  movimientos?: BienMovimiento[];
  created_at: string;
  updated_at: string;
}

export interface AssetSummaryStats {
  total_bienes: number;
  bienes_operativos: number;
  bienes_en_mantencion: number;
  bienes_de_baja: number;
  por_categoria: Record<string, number>;
}
